runtime/src/iree/hal/drivers/hip/util/tree.c

```diff
@@ -0,0 +1,582 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/hal/drivers/hip/util/tree.h"
+
+#include <stdint.h>
+
+#include "iree/base/api.h"
+
+static iree_hal_hip_util_tree_node_t*
+iree_hal_hip_util_tree_get_node_from_cache(iree_hal_hip_util_tree_t* tree) {
+  if (tree->cache) {
+    iree_hal_hip_util_tree_node_t* node = tree->cache;
+    tree->cache = node->right;
+    return node;
+  }
+  return NULL;
+}
+
+static void iree_hal_hip_util_tree_add_node_to_cache(
+    iree_hal_hip_util_tree_t* tree, iree_hal_hip_util_tree_node_t* node) {
+  node->right = tree->cache;
+  tree->cache = node;
+}
+
+static void iree_hal_hip_util_tree_delete_node(
+    iree_hal_hip_util_tree_t* tree, iree_hal_hip_util_tree_node_t* node) {
+  if (node != &tree->nil) {
+    iree_hal_hip_util_tree_add_node_to_cache(tree, node);
+  }
+}
+
+static iree_status_t iree_hal_hip_util_tree_allocate_node(
+    iree_hal_hip_util_tree_t* tree, iree_hal_hip_util_tree_node_t** out_node) {
+  iree_hal_hip_util_tree_node_t* node =
+      iree_hal_hip_util_tree_get_node_from_cache(tree);
+  if (node) {
+    memset(node, 0, sizeof(iree_hal_hip_util_tree_node_t) + tree->element_size);
+  } else {
+    IREE_RETURN_IF_ERROR(iree_allocator_malloc(
+        tree->allocator,
+        sizeof(iree_hal_hip_util_tree_node_t) + tree->element_size,
+        (void**)&node));
+  }
+  *out_node = node;
+  node->data = (uint8_t*)node + sizeof(iree_hal_hip_util_tree_node_t);
```

Comment:
when able try to use the value - as a habit this prevents things from getting out of sync (if the type of node changes and this isn't caught it will be wrong, whereas if `*node` it'll always be what you want)
```suggestion
  node->data = (uint8_t*)node + sizeof(*node);
```